README.md
# Cluster System for [tmi.js](https://github.com/tmijs/tmi.js) Twitch Bots

![License](https://img.shields.io/github/license/derpierre65/tmi.js-cluster)
[![Downloads](https://img.shields.io/npm/dt/tmi.js-cluster)](https://www.npmjs.com/package/tmi.js-cluster)
[![NPM Version](https://img.shields.io/npm/v/tmi.js-cluster)](https://www.npmjs.com/package/tmi.js-cluster)
[![Node Version](https://img.shields.io/node/v/tmi.js-cluster.svg?style=flat)](https://www.npmjs.com/package/tmi.js-cluster)
[![Issues](https://img.shields.io/github/issues/derpierre65/tmi.js-cluster)](https://github.com/derpierre65/tmi.js-cluster/issues)
[![Pull Requests](https://img.shields.io/github/issues-pr/derpierre65/tmi.js-cluster)](https://github.com/derpierre65/tmi.js-cluster/pulls)
[![Discord](https://discordapp.com/api/guilds/933758189491613707/embed.png?style=shield)](https://discord.gg/Zg4VQXZ7MG)

## Introduction

tmi.js-cluster is a scalable cluster for [tmi.js](https://github.com/tmijs/tmi.js). This cluster can have multiple supervisors that can be deployed on multiple servers.  
The cluster store its data into a database and use a redis connection for the IRC command queue to join/part channels.

## Features

- Supervisor can deployed on multiple servers.
- Use the up-to-date [tmi.js](https://github.com/tmijs/tmi.js) client.
- Monitoring dashboard.
- Optimized for **unverified** and verified bots.
- Multiple tmi clients.

## Events

| Available             | Event                | Description                                                                          | Parameters                        |
|-----------------------|----------------------|--------------------------------------------------------------------------------------|-----------------------------------|
| Supervisor            | supervisor.ready     | Supervisor is now ready.                                                             | supervisor id                     |
| Supervisor            | supervisor.error     | Supervisor couldn't spawned.                                                         | supervisor id, error              |
| Supervisor            | supervisor.terminate | Supervisor terminate started.                                                        | supervisor id                     |
| Supervisor            | supervisor.ping      | Health ping for the supervisor.                                                      | supervisor id                     |
| Supervisor            | process.create       | Process created.                                                                     | process id                        |
| Supervisor            | process.remove       | Process destroyed.                                                                   | process id                        |
| Supervisor, TmiClient | tmi.join             | Will emitted when the client join a channel. `error` is `null` if no error occurred. | error, channel                    |
| Supervisor, TmiClient | tmi.part             | Will emitted when the client part a channel. `error` is `null` if no error occurred. | error, channel                    |
| TmiClient             | tmi.channels         | Will emitted every `process.periodicTimer` ms if the tmi.js client is in open state. | array of channels (unique values) |
| TmiClient             | tmi.client.created   | Will emitted when a new client will be created (includes the main client).           | error, username, client           |
| TmiClient             | tmi.client.deleted   | Will emitted when a client will be deleted.                                          | username, client                  |

### tmi.client.created

- `error` is null if no error occurred.
- `username` is null for the main client. **It's the target channel name not the bot name.**
- `client` contains the tmi.js (or any else) client.

## Configuration

Each and every option listed below is optional.

`file`: String - The bot file that should be executed if a new process will be created.

`redis`:
- `prefix`: String - Prefix for every redis key (Default: `tmi-cluster:`)

`supervisor`:
- `keyLength`: Number - Set the key length for the supervisor id. The supervisor id will be generated from hostname and a random generated string. (Default: `8`)
- `stale`: Number - The supervisor will be marked to terminate if the last ping was more than `stale` seconds ago. (Default: `15`)
- `updateInterval`: TODO (Default: `3_000`)

`process`:
- `stale`: Number - The process will be marked to terminate if the last ping was more than `stale` seconds ago. (Default: `15`)
- `periodicTimer`: Number - After `periodicTimer` milliseconds the metrics will be saved into the database, queued commands and health check will be executed. (Default: `2_000`)
- `timeout`: Number - If the process marked to terminate, after `timeout` milliseconds the process will be killed. (Default: `60_000`)
- `terminateUncaughtException`: Boolean - If `true` the process will be terminated on uncaughtException. (Default: `true`)

`multiClients`:
- `enabled`: Boolean - If `true` then the multi client feature is enabled. You should set it to `false` if you don't use multi clients to save performance. (Default: `false`)

`metrics`:
- `enabled`: Boolean - If `true`, then metrics for every process will be generated and saved into the database. (Default: `true`)
- `memory`: Boolean - If `true`, then the current memory (MB) will be saved into the metrics object. (Default: `true`)

`autoScale`:
- `processes`:
	- `min`: Number - The minimum of processes. (Default: `2`)
	- `max`: Number - The maximum of processes. (Default: `20`)
- `thresholds`:
	- `channels`: Number - The maximum target of channels per process. (Default: `2_000`)
	- `scaleUp`: Number - If a supervisor reach more than `scaleUp`% channels then a new process will be created. (Default: `75`)
	- `scaleDown`: Number - If a supervisor reach less than `scaleDown`% channels then one process will be terminated. (Default: `50`)

`throttle`:
- `join`:
	- `allow`: Number - The maximum join/part commands in `every`ms time window. (Default: `20`)
	- `every`: Number - After this time the rate limit time window will be reset. (Default: `10_000`)
- `clients`:
	- `allow`: Number - The maximum of clients that can be created in the `every`ms time window. (Default: `50`)
	- `every`: Number - After this time the rate limit time window will be reset. (Default: `10_000`)

### Default Object

```json
{
	"file": "bot.js",
	"redis": {
		"prefix": "tmi-cluster:"
	},
	"supervisor": {
		"keyLength": 8,
		"stale": 15,
		"updateInterval": 3000
	},
	"process": {
		"stale": 15,
		"periodicTimer": 2000,
		"timeout": 60000
	},
	"multiClients": {
		"enabled": false
	},
	"metrics": {
		"enabled": true,
		"memory": true
	},
	"autoScale": {
		"processes": {
			"min": 2,
			"max": 20
		},
		"thresholds": {
			"channels": 2000,
			"scaleUp": 75,
			"scaleDown": 50
		}
	},
	"throttle": {
		"join": {
			"allow": 20,
			"every": 10000
		},
		"clients": {
			"allow": 50,
			"every": 10000
		}
	}
}
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0-alpha.3] - 2023-04-10
### Added
- **Breaking Change**: Split ChannelDistributor into ChannelDistributor (base class), RedisChannelDistributor and RedisPubSubChannelDistributor.
- `SUPERVISOR_ID` env variable for the current supervisor id.

### Changed
- **Breaking Change**: `throttle.join.every` and `throttle.clients.every` are now milliseconds instead of seconds.
- **Breaking Change**: Renamed queue redis key from `join-handler` to `command-queue`.
- Changed the way to rate limit the channel joins/parts. First join/part will start a `every`ms time window, within this window `allow` channels can be joined/parted.
- Faster unlock queue if the command queue is empty.

### Fixed
- Incorrect datetime value fixed ([#4](https://github.com/derpierre65/tmi.js-cluster/issues/4)).

### Removed
- Removed setting `throttle.join.take` and `throttle.clients.take`.

## [1.0.0-alpha.2] - 2022-09-27
### Added
- Support for multi clients.
- Option `process.terminateUncaughtException` to disable terminate on `uncaughtException`.

### Changed
- **Breaking Change**: Flipped the `tmi.join` and `tmi.part` arguments (first is the error (`null` if no error occurred), second the channel name).

### Fixed
- The order of queue commands was wrong if a client will be terminated and have an active queue in progress.

### Removed
- **Breaking Change**: Removed `joinNow` and `partNow` in ChannelDistributor. Can be replaced with the new second argument in `join` and `part`.
- Dropped `tmi.join_error` and `tmi.part_error`.

## [1.0.0-alpha.1] - 2022-07-29
### Added
- Support redis pub/sub for faster joining/parting channels.
- Priority for channel join/part order.
- Safe termination on `uncaughtException`.
- Mark a process as stale if the state is `TERMINATED` and will not waiting if the last ping is in stale range.
- Drop `PART` and `JOIN` event if a `PART` command has been found after the `JOIN` command for the given channel.
- Events (`tmi.join`, `tmi.join_error`, `tmi.part`, `tmi.part_error`) for TmiClient instance.
- Event `tmi.channels` added to TmiClient instance.

### Changed
- Updated release stale supervisors for a faster rejoin.

### Fixed
- Save current process channels if the process has been terminated to prevent loosing channels.

### Removed
- The dashboard vue.js code has been removed and moved to the laravel repository.
- Unused code.
- StaleIds from `join`, `joinNow`, `part` and `partNow`.

## [1.0.0-alpha.0] - 2022-07-08
### Added
- Initial Release

src/Supervisor.js
import merge from 'deepmerge';
import fs from 'fs';
import EventEmitter from 'node:events';
import os from 'os';
import path from 'path';
import AutoScale from './AutoScale';
import {formatDate} from './lib/date';
import * as str from './lib/string';
import ProcessPool from './ProcessPool';
import SignalListener from './SignalListener';

/**
 * @type Supervisor
 */
let SupervisorInstance = null;

class Supervisor extends EventEmitter {
	constructor(options, config) {
		super();

		this.sendMotd();

		process.env.TMI_CLUSTER_ROLE = 'supervisor';

		SupervisorInstance = this;

		const defaultConfig = {
			file: 'bot.js',
			redis: {
				prefix: 'tmi-cluster:',
			},
			supervisor: {
				keyLength: 8,
				stale: 15,
				updateInterval: 3_000,
			},
			process: {
				stale: 15,
				periodicTimer: 2_000,
				timeout: 60_000,
				terminateUncaughtException: true,
			},
			multiClients: {
				enabled: false,
			},
			metrics: {
				enabled: true,
				memory: true,
			},
			autoScale: {
				processes: {
					min: 2,
					max: 20,
				},
				thresholds: {
					channels: 2_000,
					scaleUp: 75,
					scaleDown: 50,
				},
			},
			throttle: {
				join: {
					// bot allow to join/part 20 channels every 10 seconds
					allow: 20,
					every: 10_000,
				},
				clients: {
					allow: 50,
					every: 10_000,
				},
			},
		};

		config = merge(defaultConfig, config || {});

		// force empty config object if no config given
		if (typeof options !== 'object' || options === null) {
			options = {};
		}

		// save configs
		this._config = config;
		global.tmiClusterConfig = config;

		// internal initialization
		this.id = null;
		this.database = options.database || null;

		this._promises = options.promises || {};
		this._channelDistributor = new options.channelDistributor(options);
		this._processPool = new ProcessPool();
		this._signalListener = new SignalListener(process, this);
		this._autoScale = (options.autoScale && new options.autoScale()) || new AutoScale();
		this._working = false;

		let every = config.throttle.join.every;
		if ((every - 1) * 1_000 < tmiClusterConfig.process.periodicTimer && !options.redis.sub) {
			every *= 1_000;
			console.warn(`For unverified bots its not recommended that throttle.join.every (${every}) is equal or lower then process.periodicTimer (${tmiClusterConfig.process.periodicTimer}) in non pub/sub mode.`);
		}
	}

	sendMotd() {
		const data = require('../package.json');

		console.info(fs.readFileSync(__dirname + '/motd.txt').toString());
		console.info(`You are running tmi.js-cluster v${data.version}.`);

		if (data.version.includes('alpha')) {
			console.warn('Warning: This is an alpha build. It\'s not recommended running it in production.');
		}

		console.info('');
	}

	/**
	 * @param validate Promise
	 */
	spawn(validate) {
		this.id = this.generateUniqueSupervisorId();

		if (!(validate instanceof Promise)) {
			validate = Promise.resolve();
		}

		validate
			.then(() => {
				if (!this.database) {
					return Promise.resolve();
				}

				return new Promise((resolve, reject) => {
					const now = formatDate(new Date());

					this.database.query('INSERT INTO tmi_cluster_supervisors (??) VALUES (?)', [
						['id', 'last_ping_at', 'metrics', 'options', 'created_at'],
						[this.id, now, '{}', '{}', now],
					], (error) => {
						if (error) {
							console.error('[tmi.js-cluster] The supervisor couldn\'t be created.', error);

							return reject(error);
						}

						resolve();
					});
				});
			})
			.then(() => {
				process.env.DEBUG_ENABLED && console.debug(`[tmi.js-cluster] Supervisor ${this.id} started successfully.`);

				this.emit('supervisor.ready', this.id);
				this._processPool.scale(this._config.autoScale.processes.min);

				this._working = true;
				this._interval = setInterval(async () => {
					if (this._working) {
						await this._channelDistributor.releaseStaleSupervisors();
						await this._autoScale.scale();
						await this._processPool.monitor();
						await this._channelDistributor.executeQueue();
					}

					this.database?.query('UPDATE tmi_cluster_supervisors SET last_ping_at = ? WHERE id = ?;', [
						formatDate(new Date()),
						this.id,
					], (error) => {
						if (error) {
							console.error(`[tmi.js-cluster] Supervisor ${this.id} ping update failed.`, error);
						}
					});

					this.emit('supervisor.ping', this.id);
				}, this._config.supervisor.updateInterval);
			})
			.catch((error) => {
				console.error(`[tmi.js-cluster] Supervisor ${this.id} could not be started.`, error);

				this.emit('supervisor.error', this.id, error);

				process.exit(0);
			});
	}

	terminate() {
		this._working = false;

		this.emit('supervisor.terminate', this.id);
		clearInterval(this._interval);

		process.env.DEBUG_ENABLED && console.debug(`[tmi.js-cluster] [supervisor:${this.id}] Terminating...`);

		return this
			.getPromise('terminate')
			.then(() => this._processPool.terminate())
			.then(() => this._channelDistributor.terminate())
			.then(() => {
				process.env.DEBUG_ENABLED && console.debug(`[tmi.js-cluster] [supervisor:${this.id}] Terminated.`);
				process.exit(0);
			});
	}

	getPromise(name, ...args) {
		return (this._promises[name] ? this._promises[name](...args) : Promise.resolve()) || Promise.resolve();
	}

	generateUniqueSupervisorId() {
		return os.hostname() + '-' + str.random(this._config.supervisor.keyLength);
	}

	get modulePath() {
		if (path.isAbsolute(this._config.file)) {
			return this._config.file;
		}

		return path.join(process.cwd(), this._config.file);
	}
}

export {
	SupervisorInstance,
	Supervisor as default,
};